File: src/main/java/Graph.java
import com.sun.istack.internal.NotNull;
import com.sun.istack.internal.Nullable;

import java.util.List;
import java.util.Map;
import java.util.Set;

public interface Graph {


        interface Vertex {
            @NotNull
            String getName();
        }

        interface Edge {
            @NotNull
            Vertex getBegin();

            @NotNull
            Vertex getEnd();

            default int getWeight() {
                return 1;
            }
        }

        @NotNull
        Set<Vertex> getVertices();

        @NotNull
        Set<Edge> getEdges();

        @Nullable
        Vertex get(String name);

        @NotNull
        default Set<Vertex> getNeighbors(@NotNull Vertex v) {
            return getConnections(v).keySet();
        }

        @NotNull
        Map<Vertex, Edge> getConnections(@NotNull Vertex v);

        @NotNull
        List<Vertex> getDirectedConnections(@NotNull Vertex v);

        @Nullable
        default Edge getConnection(@NotNull Vertex v1, @NotNull Vertex v2) {
            return getConnections(v1).get(v2);
        }


    }

File: src/main/java/Graph.kt
class GraphBuilder {

    data class VertexImpl(private val nameField: String) : Graph.Vertex {
        override fun getName() = nameField

        override fun toString() = name
    }

    data class EdgeImpl(
            private val weightField: Int,
            private val _begin: Graph.Vertex,
            private val _end: Graph.Vertex
    ) : Graph.Edge {
        override fun getBegin() = _begin

        override fun getEnd() = _end

        override fun getWeight() = weightField
    }

    private val vertices = mutableMapOf<String, Graph.Vertex>()

    private val connections = mutableMapOf<Graph.Vertex, Set<EdgeImpl>>()

    private fun addVertex(v: Graph.Vertex) {
        vertices[v.name] = v
    }

    fun addVertex(name: String): Graph.Vertex {
        return VertexImpl(name).apply {
            addVertex(this)
        }
    }

    fun addConnection(begin: Graph.Vertex, end: Graph.Vertex, weight: Int = 1) {
        val edge = EdgeImpl(weight, begin, end)
        connections[begin] = connections[begin]?.let { it + edge } ?: setOf(edge)
        connections[end] = connections[end]?.let { it + edge } ?: setOf(edge)
    }

    fun build(): Graph = object : Graph {

        override fun get(name: String): Graph.Vertex? = this@GraphBuilder.vertices[name]

        override fun getVertices(): Set<Graph.Vertex> = this@GraphBuilder.vertices.values.toSet()

        override fun getEdges(): Set<Graph.Edge> {
            return connections.values.flatten().toSet()
        }

        override fun getDirectedConnections(v: Graph.Vertex): List<Graph.Vertex> {
            val edges = connections[v] ?: emptySet()
            var result = mutableListOf<Graph.Vertex>()
            for (edge in edges) {
                if (v == edge.begin) {
                    result.add(edge.end)
                }
            }
            return result;
        }


        override fun getConnections(v: Graph.Vertex): Map<Graph.Vertex, Graph.Edge> {
            val edges = connections[v] ?: emptySet()
            val result = mutableMapOf<Graph.Vertex, Graph.Edge>()
            for (edge in edges) {
                when (v) {
                    edge.begin -> result[edge.end] = edge
                    edge.end -> result[edge.begin] = edge
                }
            }
            return result
        }
    }
}